RFC-0001-row-id.md

```diff
@@ -0,0 +1,80 @@
+# **RFC 1 for Presto**
+
+## Row ID
+
+Proposers
+
+* Brian Dahmen
+* Devang Shah
+* Elliotte Rusty Harold
+* Serge Druzkin
+* Sreeni Viswanadha
+
+## Related Issues
+
+* Row ID support #22078
+* Binary IDs for Hive partitions #22103
+* Binary rowID type #22053
+* Start adding row\_id hidden column to Hive #22008
+
+
+## Summary
+
+A row ID is a binary key for a row that is globally unique across the warehouse.
+It identifies a specific row at a specific version of a partition’s data. A row
+ID will continue to locate the same row in the warehouse until the underlying
+partition is mutated.
+
+## Background
+
+Queries can produce lists of row IDs that can be saved and used as input to
+future queries for quick selection. A list of row IDs can be used include or
+exclude rows to form a subset from a table.
+
+### Goals
+
+Row IDs resolve to the same value across compute engines. In this way they can
```

Comment:
It seems this goal is something a table format defines, not Presto or any other compute engine.  Hive defines the behavior for ACID transactions, which includes row IDs.  Iceberg similarly defines positions, which I think is a transferable concept.  Hudi uses primary keys, which I don't think is transferable.  My concern is we're defining concepts in the wrong place--storage defines the behavior in all the examples I listed, and Presto conforms to the spec they give.  If Presto defines it, basically it becomes esoteric Presto behavior, it's not really a spec at all.

If this is originating from Metalake, shouldn't this be something that is defined internally by Metalake?  In which case, don't you already have connectors and infrastructure that speak in terms of "metalake"?  What aspect of this proposal necessitates any change at all in core, given we already have fully functional examples of this concept purely at the connector level?

Reply:
It's not just the table format. The table format can define a row ID for rows within that table. ORC already does. However, for a global row ID across the warehouse, we need components that identify not just the row within the table but also the table (or rather the version of the partition within the table). That piece is supplied with Metastore.

I'm going to ask @kaikalur and @bmdahmen, to chime in on why this belongs in core.

Reply:
Right, but the table, the version of the table, are both things that also are connector-specific.  They may be entirely encapsulated in the connector.

Reply:
You've been working on this for longer than I have so I suspect you're seeing some approaches I'm just not thinking of. When you say encapsulate it entirely in the connector, which classes specifically are you suggesting this goes into?

By comparison, what I'm currently thinking is that Metastore stuffs the partition ID into HiveSplit. (This is roughly #22103). That HiveSplit eventually arrives as an argument in HivePageSourceProvider.createSelectivePageSource.createAggregatedPageSource
 
At this point we check to see if the split has a partition ID. If it does, we pass that partition ID down the call chain until it arrives in OrcSelectiveRecordReader.getNextPage. (code not yet written)

Now here there is already code that appends a block of row numbers if the right flag is set. I'm thinking that if a partition ID is present and if $row_id is a requested column, then here we add the block of row IDs by concatenating the row number and partition ID. I also need to stuff the row group ID in here somehow. Then it all moves back up the call stack.

There might well be some other place we can do this, and I'm just not seeing it. Can you elaborate on the alternative you'd prefer? Thanks.

Reply:
The flow you outline is all connector-specific: the Metastore is a Hive concept, as is the ORC reader (although other connectors use that part a library).

There's two parts that I'd like to think about:

1) Can the `$row_id` hidden column be added to `PrismConnector`?  If this is really hard, could we at least configure this so it's not exposed by default?  It revolves around what purpose there is around exposing row ids when you could have phantom rows.
2) Can the binary `RowIdType` itself be exposed as a new type in `PrismConnector`?  This would be a change in Prism's implementation of `Plugin#getTypes`, instead of a new type in `presto-common`.  This would encapsulate the type inside of where it's used, in `PrismConnector`.

Reply:
Looks like PrismConnector went away about 6 years ago. Do you mean HiveConnector?

Reply:
It should be subclass of the Hive connector you currently use.